### PR TITLE
Problem: #869 : Support for token price api

### LINKED ISSUE #869: Problem: CRC20 price feed is not reflectable with Coinbase API
## Problem
With our existing Coinbase API solution, CRC20 token price feed is not reflectable (or in considerable time)

#### ✅ Crypto.com Price API [https://price-api.crypto.com/price/v1/tokens](https://price-api.crypto.com/price/v1/tokens)
- Docs: https://mcoproduct.atlassian.net/wiki/spaces/C/pages/1373798547/Token+Price+REST+WebSocket+API
- There's a limit on the API avoiding the abuse of use (300 requests per minute)

Implementation: 
1. Calling https://price-api.crypto.com/price/v2/all-tokens & construct a slug list
2. Calling https://price-api.crypto.com/price/v1/tokens/[slug] based on users' assets
3. Apply current existing multi-currencies conversion

#### Crypto.org Coins API (self-maintained proxy) [https://crypto.org/api/coins/show](https://crypto.org/api/coins/show)
- Only CRO pairs are available at the moment. 
- Will need to discuss with @allthatjazzleo & @borischeuk-crypto on which pairs to add potentially. Takes development lead time.
#### CoinGecko API [https://www.coingecko.com/en/api](https://www.coingecko.com/en/api)

### PR DESCRIPTION
**Pull request by cdc-Hitesh**
Fixes #869

### CODE REVIEW

File: src/service/rpc/MarketApi.ts [hunk 1]
```diff
@@ -63,6 +55,45 @@ export class CroMarketApi implements IMarketApi {
     // throw if no price found
     throw TypeError('Could not find requested market price info from Coinbase');
   }
+
+  public async getTokenPriceFromCryptoCom(cryptoSymbol: string, fiatCurrency: string) {
+    const allTokensSlugMap: AxiosResponse<CryptoComSlugResponse[]> = await axios({
```
Comment (XinyuCRO) [inline src/service/rpc/MarketApi.ts:60]: this call does cost some time and will be called every time we request a coin's price, but the information we need from this response doesn't change much, maybe we can cache the response in memory for better performance

**Review by XinyuCRO — COMMENTED** [review]
`src/service/rpc/MarketApi.spec.ts` is too big to review, can put these test fixtures in separate files and split them from code

File: src/service/rpc/MarketApi.ts (hunk 1 above)
Comment (cdc-Hitesh) [reply src/service/rpc/MarketApi.ts:60]: 100% correct !

I originally saved it in code memory and rewrote it!
What model would you suggest to cache?

**Comment by cdc-Hitesh** [comment]
> `src/service/rpc/MarketApi.spec.ts` is too big to review, can put these test fixtures in separate files and split them from code

Fixed.

**Review by crypto-eddy — COMMENTED** [review]
Looks good to me.

Have you tried using the `crc20-backend` as the base branch to actually test how this new price API integration is working out ?

**Comment by crypto-matto** [comment]
> Looks good to me.
> 
> Have you tried using the `crc20-backend` as the base branch to actually test how this new price API integration is working out ?

I've tried merging with your `crc20-backend` branch & it works great. 

One more consideration is on the frontend side (can be a different PR), as it's likely that many CRC20 tokens are pricing at really small amount. Shall we expand to 4 decimals? Or expand to whatever it hits 4 digit of non-zeros? (I'm seeing that DeFi Wallet is doing this dynammically)
螢幕截圖 2021-12-10 上午9 57 21

**Review by crypto-matto — APPROVED** [review]

**Comment by crypto-matto** [comment]
### To be solved
- [ ] The app crashes when sending tokens which has no marketData (e.g. LP tokens)

**Review by crypto-matto — APPROVED** [review]

**Pull request merged**